lib/rate_map.c

```diff
@@ -206,3 +210,249 @@ rate_map_shift_by_mass(rate_map_t *self, double pos, double mass)
+ *
+ * `fast_search_t` is a lookup table pointing to an array of `double`s
+ * (the "elements"). It does not own the array but does point to all its memory.
+ *
+ * It's purpose is speed up the search for a "query" value in that array of
+ * elements. The `fast_search_idx_strict_upper` function is essentially a drop-in
+ * replacement for the function `idx_1st_strict_upper_bound` (which in turn is
+ * a drop-in replacement for the C++ <algorithm> library function
+ * std::upper_bound).
+ *
+ * The key concept behind `fast_search_t` is that the bits inside a
+ * "query" `double` contain a good deal of information about where near-by
+ * values in the array of elements might be found in memory. The best
+ * performance is when the array of elements have values that increase roughly
+ * linearly. But this is not a requirement. The only requirement is that the
+ * elements be non-decreasing.
+ *
+ * Since the element values are not perfectly linear some re-mapping is
+ * required.  This is achieved through an array of "lookup" pointers into the
+ * array of elements.  Although there need not be a linear mapping between
+ * element values and the range of query values, the "lookup" pointer are
+ * linear with the range of query values.
+ *
+ * The `query_multiplier` determines the linear mapping between possible query
+ * values and indexes in the array of "lookup" pointers. Because the
+ * `query_multiplier` is a power of 2, a predictable range of possible "query"
+ * values will truncate to a specific lookup index.
+ *
+ * The number of lookup pointers could be adjustable, but this simple
+ * implementation chooses one simple strategy for the number of lookup
+ * pointers. For a perfectly linear sequence of element values, the lookup
+ * pointers will be a simple element-address-by-next-element-address
+ * progression of pointers. The extremely simple case of element values being
+ * the non-negative integers is coded in the `test_fast_search_identity`
+ * unit test.  An example of alternative linear sequences with different
+ * `query_multiplier` are coded in `test_fast_search_2powers`.
+ */
+
+#if FLT_RADIX != 2
+#error "Base 2 floating point types required"
+#endif
+
+#if SIZE_MAX < ULLONG_MAX
+#error "size_t must be at least 64 bits"
+/* unsigned long long must be at least 64 bits */
+#endif
+
+static bool
+valid_sorted_nonempty_array(const double *array, size_t size)
+{
+    size_t idx;
+
+    if ((ptrdiff_t) size < 1 || isnan(array[0])) {
+        return false;
+    }
+    for (idx = 1; idx < size; idx++) {
+        if (!(array[idx - 1] <= array[idx])) {
+            return false;
+        }
+    }
+    return true;
+}
+
+static bool
+fast_search_valid(fast_search_t *self)
+{
+    const double *start, *stop;
+    size_t idx;
+
+    if (!(self->query_multiplier >= 0.0)) { // NaN not valid
+        return false;
+    }
+    if (self->num_lookups < 2) {
+        return false;
+    }
+    for (idx = 1; idx < self->num_lookups; idx++) {
```

Comment:
Thx for sharing that term nominal-case. I'll use that term in a future communication.
Good point on this function and I now see it's an unholy mix of internal state-checking and nominal-case parameter input validation. So I'll separate the two. My thinking here was an independent internal state-checking function that is carved out and not cluttering the main code and also agnostic as to how it gets used in one of four different ways:

1. terminate the process, even for end users (in which can call it inside tsk_bug_assert)
2. "if not goto out" it with MSP_ERR_ASSERTION_FAILED
3. terminate the process in only debug builds (call it from inside assert())
4. only call it directly from unit tests (and not indirectly via debug builds in automated tests)

Obviously if the function is used only via 4. then the function can exist in only the unit test code. I've been thinking of starting an initial descriptive (not prescriptive) document, maybe in FAQ format, to collect your and other's thoughts on three inter-related concepts:
* what are the debug vs release builds
* use of tsk_bug_assert vs assert
* nominal/expected/unexpected/tolerated execution scenarios

I'm hoping it can helpful to folks contributing to the C codebase.

For now I'll just opt for 4, unless you prefer opting for one of the other three. I'd probably go for BOTH 2 and 3, but that reflects my habit of being 'assert happy', never crashing the process for users and writing proprietary code that few if any people will ever read. I'm OK commiting which ever of 1 through 4 you want. However as a user it makes me very unhappy when Python libraries terminate my process.

Reply:
The way we've separated out the state-checking and input validation up to now has been to have a function ``x_verify`` on class x, e.g. [fenwick_verify](https://github.com/tskit-dev/msprime/blob/main/lib/fenwick.c#L49). We then call x_verify from the unit tests. It would be good if we could follow this pattern (so, your option 4).

We don't have any distinction between debug and release builds. There's no real point, with Python being in charge of the distribution and I'd rather not introduce this extra layer of complexity. I've tried fiddling with all sorts of flags for msprime in past and it really makes very little difference because it's fundamentally a memory-bound process. So, we build the tests with asserts turned on, and we do whatever Python does by default when we ship release binaries.

I absolutely agree 100% that Python libraries should never terminate the process. There must be 100s of millions or billions of individual msprime simulations run at this point and I would bet that < 100 of these have ended because of process termination. We have some very large number of "nines" of reliability. We are absolutely not coding for the common case here when we do a tsk_bug_assert - we're coding for the 1-in-a-hundred-billion case.

I think you're missing full context of how this C code is run: there are three layers of input validation performed before a simulation is run. Parameters are validated in the high-level Python, at the C Python layer and finally in the C code that you are working with. I feel confident in saying this is far stronger input validation that most software gets, and this frees us to write C code that is unencumbered by having to worry about all possible inputs. The inputs are *stringently* validated before the C code sees them, and this lets us write simpler, easier to understand, and more efficient C.

Reply:
> How would you classify the required sortedness of array of elements/values ... that fast_search_t "points" to?


> Should fast_search_alloc() not check and assume callers will make this validation? (rate_map_alloc currently does make this validation)

That's a bit of a grey area, since we're only using fast_search in one place, and it's already validating this requirement. I guess it's no harm checking the sortedness on the way in for fast_search_alloc, if you don't mind writing test cases to cover this.